File: frontend/src/renderer/lib/block-find.ts
import { compareMatchScores, matchRanges, scoreMatch, type MatchRange, type MatchScore } from "./text-match";
import { blockDisplay, type SessionBlock } from "./session-block";

export type BlockMatch = { blockId: string; field: "displayName" | "summary"; score: MatchScore; ranges: MatchRange[] };
export type FilterResult = { blocks: SessionBlock[]; matchIds: ReadonlySet<string>; hiddenCount: number };
export const FIND_CONTEXT_BLOCKS = 1;

export function blockSearchFields(block: SessionBlock): string[] {
	const display = blockDisplay(block);
	return [display.displayName, display.summary];
}

function matchBlock(block: SessionBlock, query: string): BlockMatch | null {
	const fields = blockSearchFields(block);
	let best: BlockMatch | null = null;
	for (const [index, field] of fields.entries()) {
		const score = scoreMatch(query, field, { subsequence: false });
		if (score === null) continue;
		const candidate: BlockMatch = {
			blockId: block.id,
			field: index === 0 ? "displayName" : "summary",
			score,
			ranges: matchRanges(query, field, score),
		};
		if (best === null || compareMatchScores(candidate.score, best.score) < 0) best = candidate;
	}
	return best;
}

export function findBlockMatches(blocks: readonly SessionBlock[], query: string): BlockMatch[] {
	if (!query.trim()) return [];
	const matches: BlockMatch[] = [];
	const visit = (block: SessionBlock): void => {
		const match = matchBlock(block, query);
		if (match) matches.push(match);
		for (const child of block.children ?? []) visit(child);
	};
	for (const block of blocks) visit(block);
	return matches;
}

export function filterBlocks(blocks: readonly SessionBlock[], query: string, contextBlocks: number): FilterResult {
	if (!query.trim()) return { blocks: blocks as SessionBlock[], matchIds: new Set(), hiddenCount: 0 };
	const matches = findBlockMatches(blocks, query);
	const matchIds = new Set(matches.map((match) => match.blockId));
	const topLevelIndex = new Map<string, number>();
	const indexBlock = (block: SessionBlock, index: number): void => {
		topLevelIndex.set(block.id, index);
		for (const child of block.children ?? []) indexBlock(child, index);
	};
	blocks.forEach(indexBlock);
	const keep = new Set<number>();
	const context = Math.max(contextBlocks, 0);
	for (const match of matches) {
		const index = topLevelIndex.get(match.blockId);
		if (index === undefined) continue;
		for (let candidate = Math.max(0, index - context); candidate <= Math.min(blocks.length - 1, index + context); candidate += 1) keep.add(candidate);
	}
	const filtered = blocks.filter((_, index) => keep.has(index));
	return { blocks: filtered, matchIds, hiddenCount: blocks.length - filtered.length };
}

export function nextMatchId(matches: readonly BlockMatch[], currentId: string | undefined, forward: boolean): string | undefined {
	if (matches.length === 0) return undefined;
	const currentIndex = currentId === undefined ? -1 : matches.findIndex((match) => match.blockId === currentId);
	if (currentIndex === -1) return forward ? matches[0]!.blockId : matches[matches.length - 1]!.blockId;
	return matches[(currentIndex + (forward ? 1 : -1) + matches.length) % matches.length]!.blockId;
}

File: packages/mobile/lib/feature/blocks/logic/block_find.dart
import 'package:equatable/equatable.dart';
import 'package:operator_mobile/core/search/text_match.dart';
import 'package:operator_mobile/feature/blocks/logic/session_block.dart';

enum BlockMatchField { displayName, summary }

class BlockMatch extends Equatable {
  const BlockMatch({required this.blockId, required this.field, required this.score, required this.ranges});

  final String blockId;
  final BlockMatchField field;
  final MatchScore score;
  final List<MatchRange> ranges;

  @override
  List<Object?> get props => [blockId, field, score, ranges];
}

class BlockFilterResult extends Equatable {
  const BlockFilterResult({required this.blocks, required this.matchIds, required this.hiddenCount});

  final List<SessionBlock> blocks;
  final Set<String> matchIds;
  final int hiddenCount;

  @override
  List<Object?> get props => [blocks, matchIds, hiddenCount];
}

sealed class BlockFind {
  static List<String> searchFields(SessionBlock block) {
    final display = blockDisplay(block);
    return [display.displayName, display.summary];
  }

  static List<BlockMatch> matches(List<SessionBlock> blocks, String query) {
    if (query.trim().isEmpty) return const [];
    final results = <BlockMatch>[];
    void visit(SessionBlock block) {
      final match = _matchBlock(block, query);
      if (match != null) results.add(match);
      for (final child in block.children ?? const <SessionBlock>[]) {
        visit(child);
      }
    }

    for (final block in blocks) {
      visit(block);
    }
    return results;
  }

  static BlockMatch? _matchBlock(SessionBlock block, String query) {
    BlockMatch? best;
    for (final entry in searchFields(block).asMap().entries) {
      final score = TextMatch.score(query, entry.value, subsequence: false);
      if (score == null) continue;
      final candidate = BlockMatch(
        blockId: block.id,
        field: entry.key == 0 ? BlockMatchField.displayName : BlockMatchField.summary,
        score: score,
        ranges: TextMatch.ranges(query, entry.value, score),
      );
      if (best == null || TextMatch.compare(candidate.score, best.score) < 0) best = candidate;
    }
    return best;
  }

  static BlockFilterResult filter(List<SessionBlock> blocks, String query, int contextBlocks) {
    if (query.trim().isEmpty) return BlockFilterResult(blocks: blocks, matchIds: const {}, hiddenCount: 0);
    final results = matches(blocks, query);
    final matchIds = results.map((match) => match.blockId).toSet();
    final topLevelIndex = <String, int>{};
    void indexBlock(SessionBlock block, int index) {
      topLevelIndex[block.id] = index;
      for (final child in block.children ?? const <SessionBlock>[]) {
        indexBlock(child, index);
      }
    }

    for (var index = 0; index < blocks.length; index += 1) {
      indexBlock(blocks[index], index);
    }
    final keep = <int>{};
    final context = contextBlocks < 0 ? 0 : contextBlocks;
    for (final match in results) {
      final index = topLevelIndex[match.blockId];
      if (index == null) continue;
      final start = (index - context) < 0 ? 0 : index - context;
      final end = (index + context) >= blocks.length ? blocks.length - 1 : index + context;
      for (var candidate = start; candidate <= end; candidate += 1) {
        keep.add(candidate);
      }
    }
    final filtered = <SessionBlock>[for (var index = 0; index < blocks.length; index += 1) if (keep.contains(index)) blocks[index]];
    return BlockFilterResult(blocks: filtered, matchIds: matchIds, hiddenCount: blocks.length - filtered.length);
  }

  static String? nextMatchId(List<BlockMatch> matches, String? currentId, {required bool forward}) {
    if (matches.isEmpty) return null;
    final currentIndex = currentId == null ? -1 : matches.indexWhere((match) => match.blockId == currentId);
    if (currentIndex == -1) return forward ? matches.first.blockId : matches.last.blockId;
    return matches[(currentIndex + (forward ? 1 : -1) + matches.length) % matches.length].blockId;
  }
}
